File: frontend/admin/src/lib/dashboard-appearance.ts
import manifest from "./dashboard-appearance-manifest.json";
import {readAppearance} from "./appearance";
export type DashboardFeature={enabled:boolean;[key:string]:any};
export type DashboardAppearance={version:1;enabled:boolean;features:Record<string,DashboardFeature>};
export const dashboardDefinitions=manifest;
export const dashboardDefaults=():DashboardAppearance=>({version:1,enabled:false,features:Object.fromEntries(manifest.map(d=>[d.key,structuredClone(d.defaults)]))});
const urlKeys=new Set(["cssUrl","image","logo","avatar"]);
const cssKeys=new Set(["size","color","shadow","position","repeat","attachment","blur","logoHeight","backToTopBottom","backToTopRight"]);
export function validateDashboard(value:DashboardAppearance):string{
 try{
  if(value.version!==1||typeof value.enabled!=="boolean")throw Error("后台美化配置版本无效");
  for(const d of manifest){
   const f=value.features[d.key];if(!f||typeof f.enabled!=="boolean")throw Error("配置缺少功能："+d.title);
   for(const key of Object.keys(f))if(!(key in d.defaults))throw Error("未知参数："+key);
   for(const [key,sample]of Object.entries(d.defaults)){
    const v=f[key];if(typeof v!==typeof sample)throw Error(d.title+"："+key+" 类型错误");
    if(typeof v==="number"&&!Number.isFinite(v))throw Error(key+" 不是有效数字");
    if(typeof v==="string"){
     if(v.length>8192||v.includes("\0"))throw Error(key+" 文本无效");
     if(v&&urlKeys.has(key)){const u=new URL(v);if(!["http:","https:"].includes(u.protocol)||u.username||u.password)throw Error(key+" 地址无效")}
     if(cssKeys.has(key)&&(/[;{}<>\n\r]/.test(v)||/url\s*\(|expression\s*\(|@import/i.test(v)))throw Error(key+" 样式值无效");
    }
   }
   for(const [key,range]of Object.entries(d.constraints) as [string,number[]][])if(f[key]<range[0]||f[key]>range[1])throw Error(d.title+"："+key+" 超出范围");
  }
  if(!Number.isInteger(value.features.effects.shatterCount))throw Error("碎片数量必须是整数");
  new Intl.DateTimeFormat("zh-CN",{timeZone:value.features.utilities.timezone});
  for(const [group,key]of [["font","size"],["appearance","blur"],["brand","logoHeight"],["effects","backToTopBottom"],["effects","backToTopRight"]])if(!/^\d+(\.\d+)?(px|rem|em|vh|vw|%)$/.test(value.features[group][key]))throw Error(key+" 需填写非负长度和单位");
  if(!["cover","contain","auto"].includes(value.features.background.size))throw Error("背景缩放方式无效");
  if(!["fixed","scroll","local"].includes(value.features.background.attachment))throw Error("背景滚动方式无效");
  if(!["no-repeat","repeat","repeat-x","repeat-y","space","round"].includes(value.features.background.repeat))throw Error("背景重复方式无效");
  return "";
 }catch(e){return String(e instanceof Error?e.message:e)}
}
export function normalizeDashboard(raw?:string|DashboardAppearance):DashboardAppearance{
 const base=dashboardDefaults();if(!raw)return base;
 try{const parsed=typeof raw==="string"?JSON.parse(raw):raw;if(parsed.version!==1||typeof parsed.enabled!=="boolean")return base;base.enabled=parsed.enabled;for(const d of manifest)if(parsed.features?.[d.key])base.features[d.key]={...base.features[d.key],...parsed.features[d.key]};return validateDashboard(base)?dashboardDefaults():base}catch{return base}
}
export function importDashboardCode(source:string):DashboardAppearance{
 const document=readAppearance(source,"NZ_DASHBOARD_CONFIG");
 if(document.configError||!document.config)throw Error(document.configError||"未找到静态 NZ_DASHBOARD_CONFIG 配置");
 const old=document.config,next=dashboardDefaults();next.enabled=true;
 for(const key of Object.keys(old))if(!manifest.some(d=>d.key===key)&&key!=="externalScripts")throw Error("存在未迁移的配置模块："+key);
 if(document.scripts.some(script=>script.enabled))throw Error("存在额外 script 标签，请先核对，原代码保持不变");
 for(const d of manifest)if(old[d.key])next.features[d.key]={...next.features[d.key],...old[d.key]};
 const recognized=new Set(["26.4.18htping.js","26.4.18shijianhuansuan.js"]);
 const scripts=old.externalScripts||[];
 for(const src of scripts){const u=new URL(src);if(!["https:","http:"].includes(u.protocol)||!recognized.has(u.pathname.split("/").pop()!))throw Error("存在未迁移的外部脚本，原始代码未修改："+src)}
 next.features.utilities.ping=scripts.some((s:string)=>s.includes("26.4.18htping.js"));
 next.features.utilities.ping0=next.features.utilities.ping;
 next.features.utilities.timeConversion=scripts.some((s:string)=>s.includes("26.4.18shijianhuansuan.js"));
 const error=validateDashboard(next);if(error)throw Error(error);return next;
}

File: cmd/dashboard/controller/dashboard_appearance.go
package controller

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/gin-gonic/gin"
	appearance "github.com/nezhahq/nezha/resource/appearance"
	"github.com/nezhahq/nezha/service/singleton"
	"io"
	"math"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

type dashboardAppearanceForm struct {
	appearanceForm
	SourceCode *string `json:"source_code"`
}

func dashboardAppearanceState() map[string]any {
	c := singleton.Conf
	config := c.DashboardAppearanceConfig
	if config == "" {
		config = `{"version":1,"enabled":false,"features":{}}`
	}
	return map[string]any{"config": json.RawMessage(config), "custom_code": c.CustomCodeDashboard, "archived_code": c.DashboardAppearanceLegacyCode, "revision": appearanceRevision(c.DashboardAppearanceConfig, c.CustomCodeDashboard+"\x00"+c.DashboardAppearanceLegacyCode)}
}
func getDashboardAppearance(c *gin.Context) (any, error) {
	settingsMutationMu.Lock()
	defer settingsMutationMu.Unlock()
	return dashboardAppearanceState(), nil
}
func validateDashboardAppearance(raw []byte) (string, error) {
	if len(raw) == 0 || len(raw) > 128<<10 {
		return "", errors.New("dashboard config size is invalid")
	}
	var doc appearanceDocument
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&doc); err != nil {
		return "", err
	}
	var trailing any
	if decoder.Decode(&trailing) != io.EOF {
		return "", errors.New("unexpected trailing JSON")
	}
	var envelope map[string]any
	_ = json.Unmarshal(raw, &envelope)
	if _, ok := envelope["enabled"].(bool); !ok || doc.Version != 1 || doc.Features == nil {
		return "", errors.New("invalid dashboard appearance schema")
	}
	var defs []appearanceDefinition
	if err := json.Unmarshal(appearance.DashboardManifest, &defs); err != nil {
		return "", err
	}
	known := map[string]bool{}
	for _, d := range defs {
		known[d.Key] = true
		f, exists := doc.Features[d.Key]
		if !exists {
			continue
		}
		if f == nil {
			return "", errors.New("invalid dashboard feature")
		}
		if _, ok := f["enabled"].(bool); !ok {
			return "", errors.New("feature enabled must be boolean")
		}
		if err := validateAppearanceValue(d.Key, f, d.Defaults); err != nil {
			return "", err
		}
		for key, limit := range d.Constraints {
			if n, ok := f[key].(float64); ok && (n < limit[0] || n > limit[1]) {
				return "", fmt.Errorf("%s is outside allowed range", key)
			}
		}
		for key, item := range f {
			s, ok := item.(string)
			if !ok {
				continue
			}
			if key == "image" || key == "avatar" || key == "cssUrl" || key == "logo" {
				if s != "" {
					u, err := url.Parse(s)
					if err != nil || u.Host == "" || u.User != nil || (u.Scheme != "http" && u.Scheme != "https") {
						return "", fmt.Errorf("%s URL is invalid", key)
					}
				}
			}
			switch key {
			case "size", "color", "shadow", "position", "repeat", "attachment", "blur", "logoHeight", "backToTopBottom", "backToTopRight":
				if strings.ContainsAny(s, ";{}<>\r\n") || regexp.MustCompile("(?i)url\\s*\\(|expression\\s*\\(|@import").MatchString(s) {
					return "", fmt.Errorf("%s CSS value is invalid", key)
				}
			}
			if key == "timezone" {
				if s == "" {
					return "", errors.New("timezone is required")
				}
				if _, err := time.LoadLocation(s); err != nil {
					return "", errors.New("invalid timezone")
				}
			}
			if (key == "size" && d.Key == "font") || key == "blur" || key == "logoHeight" || key == "backToTopBottom" || key == "backToTopRight" {
				if !regexp.MustCompile("^[0-9]+(\\.[0-9]+)?(px|rem|em|vh|vw|%)$").MatchString(s) {
					return "", fmt.Errorf("%s requires a CSS length", key)
				}
			}
		}
		if n, ok := f["shatterCount"].(float64); ok && math.Trunc(n) != n {
			return "", errors.New("shatter count must be integer")
		}
		if d.Key == "background" {
			for key, choices := range map[string]string{"size": "cover|contain|auto", "attachment": "fixed|scroll|local", "repeat": "no-repeat|repeat|repeat-x|repeat-y|space|round"} {
				if value, ok := f[key].(string); ok && !strings.Contains("|"+choices+"|", "|"+value+"|") {
					return "", fmt.Errorf("invalid background %s", key)
				}
			}
		}
	}
	for key := range doc.Features {
		if !known[key] {
			return "", fmt.Errorf("unknown dashboard feature %s", key)
		}
	}
	normalized, err := json.Marshal(doc)
	return string(normalized), err
}
func updateDashboardAppearance(c *gin.Context) (any, error) {
	c.Request.Body = http.MaxBytesReader(c.Writer, c.Request.Body, 2<<20)
	var form dashboardAppearanceForm
	if err := c.ShouldBindJSON(&form); err != nil {
		return nil, err
	}
	next, err := validateDashboardAppearance(form.Config)
	if err != nil {
		return nil, err
	}
	settingsMutationMu.Lock()
	defer settingsMutationMu.Unlock()
	conf := singleton.Conf
	err = saveDashboardAppearance(&conf.DashboardAppearanceConfig, &conf.CustomCodeDashboard, &conf.DashboardAppearanceLegacyCode, form, next, conf.Save)
	if err != nil {
		return nil, err
	}
	return dashboardAppearanceState(), nil
}
func saveDashboardAppearance(config, code, archive *string, form dashboardAppearanceForm, next string, save func() error) error {
	if form.Revision != appearanceRevision(*config, *code+"\x00"+*archive) {
		return errors.New("dashboard appearance changed; reload settings")
	}
	if (form.ExpectedCustomCode == nil) != (form.RemainingCustomCode == nil) {
		return errors.New("both migration fields are required")
	}
	nextCode, nextArchive := *code, *archive
	if form.ExpectedCustomCode != nil {
		if *form.ExpectedCustomCode != *code {
			return errors.New("dashboard custom code changed")
		}
		if len(*form.RemainingCustomCode) > 1<<20 {
			return errors.New("custom code too large")
		}
		nextCode = *form.RemainingCustomCode
		if nextCode != *code {
			if *archive != "" && *archive != *code {
				return errors.New("refusing to overwrite archived source")
			}
			nextArchive = *code
		}
	}
	if form.SourceCode != nil && *form.SourceCode != "" {
		if len(*form.SourceCode) > 1<<20 {
			return errors.New("source code too large")
		}
		if *code != "" && *form.SourceCode != *code {
			return errors.New("imported source differs from current custom code")
		}
		if nextArchive != "" && nextArchive != *form.SourceCode {
			return errors.New("refusing to overwrite archived source")
		}
		nextArchive = *form.SourceCode
	}
	if *config == next && *code == nextCode && *archive == nextArchive {
		return nil
	}
	oldConfig, oldCode, oldArchive := *config, *code, *archive
	*config, *code, *archive = next, nextCode, nextArchive
	if err := save(); err != nil {
		*config, *code, *archive = oldConfig, oldCode, oldArchive
		return err
	}
	return nil
}
